differs/package_differs.go

```diff
@@ -98,3 +141,39 @@ func singleVersionAnalysis(image pkgutil.Image, analyzer SingleVersionPackageAna
 	}
 	return &analysis, nil
 }
+
+// singleVersionLayerAnalysis returns the packages included, deleted or
+// updated in each layer
+func singleVersionLayerAnalysis(image pkgutil.Image, analyzer SingleVersionPackageLayerAnalyzer) (*util.SingleVersionPackageLayerAnalyzeResult, error) {
+	pack, err := analyzer.getPackages(image)
+	if err != nil {
+		return &util.SingleVersionPackageLayerAnalyzeResult{}, err
+	}
+	var pkgDiffs []util.PackageDiff
+
+	// Each layer with modified packages includes a complete list of packages
+	// in its package database. Thus we diff the current layer with the
+	// previous one. Not all layers may include differences in packages, those
+	// are omitted.
+	preInd := -1
+	for i := range pack {
+		var pkgDiff util.PackageDiff
+		if preInd < 0 && len(pack[i]) > 0 {
+			pkgDiff = util.GetMapDiff(make(map[string]util.PackageInfo), pack[i])
+			preInd = i
+		} else if preInd >= 0 && len(pack[i]) > 0 {
+			pkgDiff = util.GetMapDiff(pack[preInd], pack[i])
+			preInd = i
+		}
+
+		pkgDiffs = append(pkgDiffs, pkgDiff)
+	}
```

Comment:
> Each layer with modified packages includes a complete list of packages in its package database. Thus we diff the current layer with the previous one.

Could you explain this a bit more? Are you saying that each layer with modified packages contains all packages installed in the previous layer, so diffing those two gives you only the packages modified between layer `i` and layer `i-1`?

I think this block would also be a little more clear if it looked something like 
```golang
for i := range pack {
	if len(pack[i] == 0) {
		continue
	}
	if i == 0 {
		pkgDiffs = append(pkgDiffs, util.GetMapDiff(make(map[string]util.PackageInfo), pack[i]))
	} else {
		pkgDiffs = append(pkgDiffs, util.GetMapDiff(pack[i-1], pack[i]))
	}
}
```

Reply:
Right, I haven't coded it as you suggest because we can't expect to have different packages in each layer. Image there are some layers that do not modify the package database, because they configure files or run some scripts or whatever; for those layers the reuslt of the getPackages is an empty map. I don't want to diff those empty maps with any other layer that actually modifies the package database, because the result will be the same as treating any package (regardless the layer in which they were appended) as a new addition.

So what I do here to compare only layers which actually include a modified package database. That means the diff is not with the immediate previous layer, but with the previous layer that actually includes some package database.

Reply:
Ahh ok, that makes sense now. Could you just update your comment slightly to explain that it's not the previous layer, but the previous layer that contains a package db?